w3rc.go

```diff
@@ -0,0 +1,56 @@
+package w3rc
+
+import (
+	"context"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+)
+
+type config struct{}
+
+// An Option allows opening a Session with configured options.
+type Option func(config) error
+
+// NewSession creates a Session with given configuration.
+// A session represents a set of related queries for content addressed data.
+// Connections to peers may stay open for the life of a session.
+func NewSession(ls ipld.LinkSystem, opts ...Option) Session {
+	return nil
+}
+
+// ResultChan provides progress updates from a call to `GetStream`
+type ResultChan chan ProgressResult
+
+// A ProgressResult is an individual update from a call to `GetStream`
+// The result will either have a status of `Error` and an Error set,
+// or will have a node and path set.
+// The ResultChan a result is sent down will close after an error result
+// or 'complete' result is sent.
+type ProgressResult struct {
+	Status
+	Error error
+	ipld.Path
+	ipld.Node
+}
+
+// Status is a code of the type of an individual Progress Result
+type Status uint8
+
+// These are valid status codes
+const (
+	ERROR Status = iota
+	INPROGRESS
+	COMPLETE
+)
+
+// A Session is able to fetch content addressed data.
+type Session interface {
+	// Get returns a dag rooted at root. If selector is `nil`, the single block
+	// of the root will be assumed. If the full dag under root is desired, (following links)
+	// `CommonSelector_MatchAllRecursively` should be provided.
+	Get(ctx context.Context, root cid.Cid, selector selector.Selector) (ipld.Node, error)
+	GetStream(ctx context.Context, root cid.Cid, selector selector.Selector) ResultChan
```

Comment:
So further to that selector comment above - what is the `ipld.Node` being returned for each `ProgressResult`? Is it each _block_ that is found during a traversal, or each _node_ that is matched by the selector? So `MatchAllRecursively` would then return every single node within every single block that's found in the DAG since it's matching absolutely everything. But then `ExploreAllRecursively` isn't actually matching anything, it's just exploring the DAG for the purpose of finding all the blocks.

Maybe this should be clear that matching nodes isn't going to matter during the traversal, it's the entire blocks that you're getting on each result as those blocks are discovered during a traversal. Hence `MatchAllRecursively` and `ExploreAllRecursively` will have the same result here, but the latter will be more efficient since it won't be calling the `visitFn` callback on each node within each block.

Reply:
my intention was that get was a simpler way to get started that could be used in the common case (where you don't care about tracking progress) and then a more expressive variant that gives you additional control, where get would be a subset of getNodes that just waits until the final value